primitives/src/vstaging/executor_params.rs

```diff
@@ -0,0 +1,151 @@
+pub enum ExecutionEnvironment {
+	/// Generic Wasmtime executor
+	WasmtimeGeneric = 0,
+}
+
+/// A single executor parameter
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, TypeInfo)]
+pub enum ExecutorParam {
+	/// ## General parameters:
+	/// Execution environment type
+	Environment(ExecutionEnvironment),
+
+	/// ## Parameters setting the executuion environment semantics:
+	/// Max. memory size
+	MaxMemorySize(u32),
+	/// Wasm logical stack size limit (max. number of Wasm values on stack)
+	StackLogicalMax(u32),
+	/// Executor machine stack size limit, in bytes
+	StackNativeMax(u32),
+
+	/// Placeholder for any data not fitting into the semantics. It may be used to avoid
+	/// `SessionInfo` migration if fast adoption of a new parameter is needed. This is an
+	/// emergency measure, only use it if you ought to.
+	#[codec(index = 255)]
+	RawData(Vec<u8>),
+}
+
+impl ExecutorParam {
+	fn is_environment(&self) -> bool {
+		match self {
+			ExecutorParam::Environment(_) => true,
+			_ => false,
+		}
+	}
+}
+
+/// Unit type wrapper around [`type@Hash`] that represents an execution parameter set hash.
+///
+/// This type is produced by [`ExecutorParams::hash`].
+#[derive(Clone, Copy, Encode, Decode, Hash, Eq, PartialEq, PartialOrd, Ord, TypeInfo)]
+pub struct ExecutorParamsHash(Hash);
+
+impl ExecutorParamsHash {
+	/// Create a new executor parameter hash from `H256` hash
+	pub fn from_hash(hash: Hash) -> Self {
+		Self(hash)
+	}
+}
+
+impl sp_std::fmt::Display for ExecutorParamsHash {
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter<'_>) -> sp_std::fmt::Result {
+		self.0.fmt(f)
+	}
+}
+
+impl sp_std::fmt::Debug for ExecutorParamsHash {
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter<'_>) -> sp_std::fmt::Result {
+		write!(f, "{:?}", self.0)
+	}
+}
+
+impl sp_std::fmt::LowerHex for ExecutorParamsHash {
+	fn fmt(&self, f: &mut sp_std::fmt::Formatter<'_>) -> sp_std::fmt::Result {
+		sp_std::fmt::LowerHex::fmt(&self.0, f)
+	}
+}
+
+/// # Deterministically serialized execution environment semantics
+/// Represents an arbitrary semantics of an arbitrary execution environment, so should be kept as
+/// abstract as possible.
+// ADR: For mandatory entries (like `ExecutorParam::Environment`), mandatoriness should be enforced
+// in code rather than separating them into individual fields of the structure. Thus, complex
+// migrations shall be avoided when adding new entries and removing old ones.
+#[derive(Clone, Debug, Encode, Decode, PartialEq, Eq, TypeInfo)]
+pub struct ExecutorParams(Vec<ExecutorParam>);
+
+impl ExecutorParams {
+	/// Creates a new, empty executor parameter set
+	pub fn new(environment: ExecutionEnvironment) -> Self {
+		ExecutorParams(vec![ExecutorParam::Environment(environment)])
+	}
+
+	/// Returns execution environment type
+	pub fn environment(&self) -> ExecutionEnvironment {
+		debug_assert!(self.0.len() > 0);
+		if let ExecutorParam::Environment(environment) = self.0[0] {
+			environment
+		} else {
+			unreachable!();
```

Comment:
```suggestion
			unreachable!("`ExecutorParam::Environment` is required to be the first parameter");
```

Something like this.